### PR TITLE
Remove direct access to governance.ChainConfig

### PR DESCRIPTION
**Pull request by andybclee**
## Proposed changes

- This PR has following changes and formerly reviewed at https://github.com/ground-x/klaytn/pull/3682
- Added getters for each filed of chain config
- Replace direct access to chainConfig with newly added getters
- From this PR, ChainConfig is not updated. governance.currentSet holds all current governance information and provides it if requested.
- blockchain has its own chainConfig and It should be updated when there is a change. A getter and setter for `UseGiniCoeff` and `ProposerPolicy` are added and those are called by governance when there is a change

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: consensus/istanbul/validator/weighted.go [hunk 1]
```diff
@@ -554,7 +554,7 @@ func (valSet *weightedCouncil) Policy() istanbul.ProposerPolicy { return valSet.
 // It return no error when weightedCouncil
 //   (1) already has up-do-date proposers
 //   (2) successfully calculated up-do-date proposers
-func (valSet *weightedCouncil) Refresh(hash common.Hash, blockNum uint64, config *params.ChainConfig) error {
+func (valSet *weightedCouncil) Refresh(hash common.Hash, blockNum uint64, chainId uint64) error {
```
Comment (jimni1222) [inline consensus/istanbul/validator/weighted.go]: When is chainId used?

File: contracts/reward/reward_test.go [hunk 2]
```diff
@@ -149,53 +151,72 @@ func TestUpdateGovernanceParameterByEpoch(t *testing.T) {
...
 		t.Errorf("GovernanceParameter is different")
 	}
 
+	blockNumber = 31
+	currentSet := governance.NewGovernanceSet()
+	delta := governance.NewGovernanceSet()
+	_, current, _ := gov.ReadGovernance(blockNumber)
+	currentSet.Import(current)
+	delta.SetValue(params.Ratio, "40/30/30")
+	gov.WriteGovernance(blockNumber-1, currentSet, delta)
```
Comment (jimni1222) [inline contracts/reward/reward_test.go]: This also just question.
Why 'blockNumber-1' as a parameter to gov.WriteGovernance?

File: consensus/istanbul/validator/weighted.go (hunk 1 above)
Comment (laggu) [reply consensus/istanbul/validator/weighted.go]: chainId was for hardfork. It was used for checking if the network is baobab.
chainId will be removed in the next reward refactoring pr

File: governance/handler.go [hunk 3]
```diff
@@ -44,67 +44,58 @@ var (
...
-	params.UseGiniCoeff:            {boolT, checkUint64andBool, updateGovernanceConfig},
-	params.DeferredTxFee:           {boolT, checkUint64andBool, updateGovernanceConfig},
-	params.MinimumStake:            {stringT, checkBigInt, updateGovernanceConfig},
-	params.StakeUpdateInterval:     {uint64T, checkUint64andBool, updateGovernanceConfig},
-	params.ProposerRefreshInterval: {uint64T, checkUint64andBool, updateGovernanceConfig},
-	params.Epoch:                   {uint64T, checkUint64andBool, updateGovernanceConfig},
-	params.Policy:                  {uint64T, checkUint64andBool, updateGovernanceConfig},
-	params.CommitteeSize:           {uint64T, checkUint64andBool, updateGovernanceConfig},
-	params.ConstTxGasHumanReadable: {uint64T, checkUint64andBool, updateParams},
+	params.GovernanceMode:          {stringT, checkGovernanceMode, nil},
```
Comment (jimni1222) [inline governance/handler.go]: Is there a criterion for when I need to handle a trigger?

File: blockchain/blockchain.go [hunk 4]
```diff
@@ -101,8 +101,9 @@ type CacheConfig struct {
 // included in the canonical one where as GetBlockByNumber always represents the
 // canonical chain.
 type BlockChain struct {
-	chainConfig *params.ChainConfig // Chain & network configuration
-	cacheConfig *CacheConfig        // stateDB caching and trie caching/pruning configuration
+	chainConfig   *params.ChainConfig // Chain & network configuration
+	chainConfigMu *sync.RWMutex
```
[the remark this answers is not in the document]
Comment (KimKyungup) [reply blockchain/blockchain.go]: `chainConfig` is a pointer. So to protect `chainConfig`, this code is better.

[the remark this answers is not in the document]
Comment (KimKyungup) [reply blockchain/blockchain.go]: But `chainConfig` didn't change, `chainConfig` can have the Mu. :)

File: consensus/istanbul/validator.go [hunk 5]
```diff
@@ -93,7 +92,7 @@ type ValidatorSet interface {
 	IsSubSet() bool
 
 	// Refreshes a list of candidate proposers with given hash and blockNum
-	Refresh(hash common.Hash, blockNum uint64, config *params.ChainConfig) error
+	Refresh(hash common.Hash, blockNum uint64, chainId uint64) error
```
Comment (KimKyungup) [inline consensus/istanbul/validator.go]: Where is the implementation of this method which needs chainID?

Comment (KimKyungup) [reply consensus/istanbul/validator.go]: https://github.com/klaytn/klaytn/pull/97/files#r309090594

File: contracts/reward/reward_test.go (hunk 2 above)
Comment (KimKyungup) [reply contracts/reward/reward_test.go]: I also wondering this. :) Could you write the reason? :)

File: blockchain/blockchain.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (andybclee) [reply blockchain/blockchain.go]: @ehnuje Changing params.ChainConfig will cause a hard fork. The struct itself is stored in a database, so to change it, all nodes have to be updated. Just for adding a lock, I think it costs too much.

[the remark this answers is not in the document]
Comment (ehnuje) [reply blockchain/blockchain.go]: @andybclee I understood, thanks for the comment!

File: governance/handler.go [hunk 6]
```diff
@@ -44,67 +44,58 @@ var (
...
+}
+
+func updateStakingUpdateInterval(g *Governance, k string, v interface{}) bool {
+	params.SetStakingUpdateInterval(g.StakingUpdateInterval())
+	return true
+}
+
+func updateProposerUpdateInterval(g *Governance, k string, v interface{}) bool {
+	params.SetProposerUpdateInterval(g.ProposerUpdateInterval())
+	return true
```
[the remark this answers is not in the document]
Comment (andybclee) [reply governance/handler.go]: @ehnuje Initially, I thought there can be some error case while reflecting changes. So the struct is defined as below.
```go
type check struct {
	t         reflect.Type
	validator func(k string, v interface{}) bool
	trigger   func(g *Governance, k string, v interface{}) bool
}
```
The above function and others which are used as trigger should have same signature. 

But at the moment, all changes are about setting a value and nothing else. So the return value is practically meaningless.

For now, I think removing return value is okay. I'll commit the change soon.

File: governance/handler.go (hunk 3 above)
Comment (andybclee) [reply governance/handler.go]: @jimni1222 This map defines 
1. Governance Item (as a key of the map)
2. the key's data type
3. the validator function for the type
4. trigger function which to be called when the value of the key changed if it requires more action other than storing it in `governance.currentSet`

So it is triggered when the change have to affect an object other than `governance`

Comment (jimni1222) [reply governance/handler.go]: Ah, ! that means, if update need to trigger something more than action in governance.currentSet, then trigger should be defined ! Thank you for explain :)

**Pull request merged**